File: internal/eval/bucket.go
package eval

import (
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"math"
	"strconv"

	"pennant/internal/model"
)

// BucketScale = 2^60 - 1. We take 15 hex chars (60 bits) from SHA-1 digest.
const BucketScale = 0xFFFFFFFFFFFFFFF

// ComputeBucket returns a deterministic float in [0.0, 1.0) for this context.
// Properties: deterministic, uniform, independent per flag, monotone under growth.
// Uses SHA-1 (not MurmurHash) because it's in every language's stdlib with
// byte-identical output — critical for cross-language parity.
func ComputeBucket(ctx *model.Context, bucketBy, key, salt string, seed *int) float64 {
	if bucketBy == "" {
		bucketBy = "key"
	}

	idValue, ok := ctx.GetAttribute(bucketBy)
	if !ok {
		return 0.0
	}

	idStr, ok := stringifyBucketValue(idValue)
	if !ok {
		return 0.0
	}

	var hashInput string
	if seed != nil {
		hashInput = fmt.Sprintf("%d.%s", *seed, idStr)
	} else {
		hashInput = fmt.Sprintf("%s.%s.%s", key, salt, idStr)
	}

	sum := sha1.Sum([]byte(hashInput))
	hexPrefix := hex.EncodeToString(sum[:])[:15]
	n, err := strconv.ParseInt(hexPrefix, 16, 64)
	if err != nil {
		return 0.0
	}
	return float64(n) / float64(BucketScale)
}

// stringifyBucketValue: ONLY strings and integers are valid bucketing values.
// Cross-language parity trap: float and bool string representations differ
// between languages. Integral floats are allowed (JS numbers are all floats).
func stringifyBucketValue(v any) (string, bool) {
	switch t := v.(type) {
	case string:
		return t, true
	case int:
		return strconv.Itoa(t), true
	case int64:
		return strconv.FormatInt(t, 10), true
	case float64:
		if t == math.Trunc(t) && math.Abs(t) < 1e15 {
			return strconv.FormatInt(int64(t), 10), true
		}
		return "", false
	default:
		return "", false
	}
}

File: sdk/ts/src/bucket.ts
import { createHash } from 'crypto'
import type { EvalContext } from './types.js'

// BucketScale is 2^60 - 1 (15 hex Fs = 0xFFFFFFFFFFFFFFF).
// We MUST use BigInt here because 60-bit integers exceed Number.MAX_SAFE_INTEGER
// (2^53 - 1 = 9007199254740991). Using a plain JS number would silently lose
// precision and produce different bucket values from the Go reference implementation.
const BUCKET_SCALE = BigInt('0xFFFFFFFFFFFFFFF')
// Pre-compute the divisor as a number for the final division step.
// Number(BUCKET_SCALE) is exact because floats can represent this integer precisely
// when used only as the denominator — the loss of precision in integer representation
// does not affect the ratio when the numerator is also subject to the same conversion.
const BUCKET_SCALE_NUM = Number(BUCKET_SCALE)

// computeBucket returns a deterministic float in [0.0, 1.0) for the given context.
//
// Hash input format (MUST match Go implementation exactly):
//   - Without seed: "{flagKey}.{salt}.{idStr}"
//   - With seed:    "{seed}.{idStr}"
//
// We take the first 15 hex characters of the SHA-1 digest (= 60 bits),
// parse them as a big integer, and divide by BucketScale.
export function computeBucket(
  ctx: EvalContext,
  bucketBy: string,
  flagKey: string,
  salt: string,
  seed?: number,
): number {
  if (!bucketBy) bucketBy = 'key'

  const attrVal = getAttribute(ctx, bucketBy)
  if (attrVal === undefined || attrVal === null) return 0.0

  const strVal = stringifyBucketValue(attrVal)
  if (strVal === null) return 0.0

  // Hash input must match Go fmt.Sprintf exactly.
  let input: string
  if (seed !== undefined && seed !== null) {
    // With seed: "{seed}.{idStr}"
    input = `${seed}.${strVal}`
  } else {
    // Without seed: "{flagKey}.{salt}.{idStr}"
    input = `${flagKey}.${salt}.${strVal}`
  }

  const hash = createHash('sha1').update(input, 'utf8').digest('hex')

  // Take the first 15 hex characters = 60 bits
  const prefix = hash.slice(0, 15)
  const big = BigInt(`0x${prefix}`)

  return Number(big) / BUCKET_SCALE_NUM
}

function getAttribute(ctx: EvalContext, name: string): unknown {
  switch (name) {
    case 'key': return ctx.key
    case 'kind': return ctx.kind ?? 'user'
    case 'anonymous': return ctx.anonymous ?? false
    default: return ctx.attributes?.[name]
  }
}

// stringifyBucketValue converts an attribute value to the string used for hashing.
// ONLY strings and integral numbers are valid. Non-integral floats return null.
// This matches the Go stringifyBucketValue function exactly.
function stringifyBucketValue(v: unknown): string | null {
  if (typeof v === 'string') return v
  if (typeof v === 'number') {
    // Integral floats are allowed (JS numbers are all floats).
    // Match Go's condition: t == math.Trunc(t) && math.Abs(t) < 1e15
    if (Number.isInteger(v) && Math.abs(v) < 1e15) {
      return String(Math.trunc(v))
    }
    return null
  }
  return null
}
